### ISSUE TITLE
MSUnmerged logic flaw or how to speed it up

### ISSUE THREAD

**Issue by vkuznet** [dmwm/WMCore#10707]
**Impact of the new feature**
A brief look at MSUnmerged codebase reveals a potential bottleneck which may significantly slow down MSUnmerged clean-up step.

**Is your feature request related to a problem? Please describe.**
If you'll closely look at these lines
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py#L270-L288
then you'll see that it is doing **sequential** loop over all files which should be cleaned up by calling rucio getPFN API. If we have large amount of files to clean-up and account for remote getPFN API time, then the total time of this loop will be be significant. For instance, getPFN takes 0.1 sec and we loop over 10K files, the total time of this loop will be 16 minutes. But if you perform all calls in parallel then the total time should be in a 1 min range.

**Describe the solution you'd like**
To resolve this and similar issues in MSUnmerged codebase the code should parallelize calls to rucio and all clean-up actions. Since I fixed similar issues in DAS, Unified, and other codes in past I expect that the gain will be significant.

**Describe alternatives you've considered**
The MSUnmerged codebase is a good candidate of possible migration to Go based code since it will simplify all concurrent tasks and speed it up significantly. To make further concrete suggestion I suggest to write down the logic of MSUnmerged. In my view it is as simple as:
- get list of files
- resolve LFN to PFNs
- perform clean-up step
The last two items can be done in parallel rather than sequentially (as it is done now).

**Additional context**

**Comment by vkuznet** [comment]
FYI: @todor-ivanov , @amaltaro , @klannon , @KatyEllis

**Comment by todor-ivanov** [comment]
Hi @vkuznet, Thanks for opening this issue, but I think this is not an issue at all, and I am about to close it with the current comment. The reasons I think so are the following:
 
* First, If you read the code carefully you will see that  you are referring here to a fall back mechanism, rather than the default behavior of the code.  In the original mechanism the `pfn prefix` is created much earlier in the code [1], so the `lfn` to `pfn` resolution is done only once per RSE, but not once per file. And it takes one and only one file from the list [2] instead of every one of them. The reason for this fallback mechanism to exist is one particular site - DESY, where in the original `pfn` creation at the site a regular expression is used instead of a well defined end path. Here is an example of a normal RSE:
```
In [14]: rcl.lfns2pfns('T2_US_Wisconsin', ['cms://store/unmerged/Run2016E/DoubleEG/MINIAOD/HIPM_UL2016_MiniAODv2-v1/120000/13756C1C-A9AF-624E-B370-6EC5BE62279E.root'], operation='delete')
Out[14]: {'cms://store/unmerged/Run2016E/DoubleEG/MINIAOD/HIPM_UL2016_MiniAODv2-v1/120000/13756C1C-A9AF-624E-B370-6EC5BE62279E.root': 'gsiftp://cms-lvs-gridftp.hep.wisc.edu:2811/hdfs/store/unmerged/Run2016E/DoubleEG/MINIAOD/HIPM_UL2016_MiniAODv2-v1/120000/13756C1C-A9AF-624E-B370-6EC5BE62279E.root'}
```
 
And here is the same resolution at DESY:
```
In [15]: rcl.lfns2pfns('T2_DE_DESY', ['cms://store/unmerged/Run2016E/DoubleEG/MINIAOD/HIPM_UL2016_MiniAODv2-v1/120000/13756C1C-A9AF-624E-B370-6EC5BE62279E.root'], operation='delete')
Out[15]: {'cms://store/unmerged/Run2016E/DoubleEG/MINIAOD/HIPM_UL2016_MiniAODv2-v1/120000/13756C1C-A9AF-624E-B370-6EC5BE62279E.root': 'srm://dcache-se-cms.desy.de:8443/srm/managerv2?SFN=/pnfs/desy.de/cms/tier2/unmerged/Run2016E/DoubleEG/MINIAOD/HIPM_UL2016_MiniAODv2-v1/120000/13756C1C-A9AF-624E-B370-6EC5BE62279E.root'}
```

**Notice** the `/store/` part cut from the prefix, which basically breaks the checks for validity of the so found path later. And since we have no idea about what would be the regular expression used by the site, I needed to have this fallback mechanism, and I needed to make it in the cheapest possible way, which I explain how I achieved it in the next bullet.

 * Second, this is done not per every file but rather per every directory from the level/branch of the `protectedLFN` list, which means basically once per workflow, which would be not that much even for a heavily loaded site, which has fallen in the quite unfortunate situation to use the fallback mechanism. BTW so far we have not noticed any other site to use it besides DESY. There could be others too, but I doubt they are too many.
 
All those warnings and concerns you mention here I have taken into account from the very beginning and they are not overlooked. I can assure you the way how this service is designed is with the thought of the cheapest possible mechanisms in terms of calls to external systems and especially to Rucio. 

p.s. Sorry for my late reply, but I was on vacation, and I see your message just now. 


[1]

https://github.com/dmwm/WMCore/blob/8c73e7073fd8928326884e3d32fd9c806dcb7429/src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py#L123

https://github.com/dmwm/WMCore/blob/8c73e7073fd8928326884e3d32fd9c806dcb7429/src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py#L555-L577

[2]
https://github.com/dmwm/WMCore/blob/8c73e7073fd8928326884e3d32fd9c806dcb7429/src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py#L567

**Comment by todor-ivanov** [comment]
As of parallelization of the last two steps you are talking about:
* resolve LFN to PFNs
* perform clean-up step

As I mentioned in my previous comment the resolution step is quite rare (despite the initial impression while one looks at the code) and needs no parallelization in y opinion. While the later - the cleanup step, it suffers from the the constraints on how the `gfal2` client works. It is impossible to do anything else but a single `gfal rm` call per file. The most I could do was to send calls of bulk deletions and let the `gfal2` client deal with the list of files I provide in the pest possible way it has driven by its own internals. For the reason I create the list generators for those file lists on a per directory level and what I did was to climb as high as possible in the `lfn` tree and start aggregating these lists from there down. In our case this is again the `protectedLFN` list level. 

As of the service parallelization in general (talking about threads and stuff here, not about parallel calls to external services), we have decided to use the thread creation mechanisms provided by default by `MSManager` and upon that to create different instances of the service using the Kubernetes machinery, and the saga there is a different story.
